src/components/TreeGrid.ts

```diff
@@ -46,13 +46,43 @@ const wrapContent =
     state: 'closed',
   });
 
+const hierarchyMapToRows = <T extends RowContents>(hierarchyMap: Map<number, T[]>, domainOptionRoot: T): Row<T>[] => {
+  const traverseHierarchy = (parent: T, depth: number, previousRows: Row<T>[]): Row<T>[] => {
+    // does parent have children?
+    const children = hierarchyMap.get(parent.id) || [];
+    const parentRow: Row<T> = {
+      contents: parent,
+      depth,
+      isFetched: children.length > 0,
+      state: children.length > 0 ? 'open' : 'closed',
+    };
+    // recursively traverse hierarchy of all children
+    const childRows: Row<T>[] = children.flatMap((child) => traverseHierarchy(child, depth + 1, []));
+    return [...previousRows, parentRow, ...childRows];
+  };
+
+  const rows: Row<T>[] = [];
+  const depth = 0;
+  // get all children of domain option root
+  const domainOptionRootChildren = hierarchyMap.get(domainOptionRoot.id) || [];
+
+  domainOptionRootChildren.forEach((child) => {
+    rows.push(...traverseHierarchy(child, depth, []));
+  });
+
+  return rows;
```

Comment:
Is this necessary? I would expect something like this to work:
```suggestion
  return traverseHierarchy(domainOptionRoot, 0, []);
```

As before I'd like to avoid mutability if possible

Reply:
Yes! You're absolutely right!